File: src/forge/agents/test_agent.py
import json
import subprocess
from pathlib import Path
from forge.agents.base import BaseAgent, AgentResult, _extract_json
from forge.router import ModelTier

SYSTEM = """You are a test engineer. Write tests for the project shown in the workspace.

Output ONLY a JSON array of test files:
[{"path": "tests/test_component.jsx", "content": "full test code"}, ...]

Critical rules:
- ONLY import modules that ACTUALLY EXIST in the workspace file tree provided
- Do NOT invent utility functions (addNumbers, runApp, etc.) — test what is actually there
- For React+Vitest: import components from their real paths (e.g. '../src/App.jsx')
- For React components: use @testing-library/react + jsdom; add `@vitest/browser` or `environment: 'jsdom'` in vitest config if needed
- For vitest: `import { describe, it, expect, vi } from 'vitest'`
- Keep tests simple — render the component and assert it mounts without crashing
- Skip a module if you cannot identify what to import from the workspace
- Use the test framework specified in the architecture"""

FRAMEWORK_CMD: dict[str, list[str]] = {
    "pytest": ["python", "-m", "pytest", "-v", "--tb=short"],
    "vitest": ["npx", "vitest", "run"],
    "jest": ["npx", "jest", "--no-coverage"],
    "react-scripts": ["npx", "react-scripts", "test", "--watchAll=false", "--passWithNoTests"],
    "go-test": ["go", "test", "./..."],
}


class TestAgent(BaseAgent):
    tier = ModelTier.REASONING

    async def run(self, workspace: Path, architecture: str) -> AgentResult:  # type: ignore[override]
        arch = json.loads(architecture) if isinstance(architecture, str) else architecture
        framework = arch.get("test_framework", "pytest")

        from forge.agents.integration import _read_workspace
        tree = _read_workspace(workspace)
        messages = [
            {"role": "system", "content": SYSTEM},
            {"role": "user", "content": f"Framework: {framework}\n\nWorkspace:\n{tree}"},
        ]
        response = await self._call(messages)
        try:
            test_files: list[dict] = json.loads(_extract_json(response))
        except json.JSONDecodeError:
            return AgentResult(success=False, output=response, error="invalid_json")

        for tf in test_files:
            path = workspace / tf["path"]
            path.parent.mkdir(parents=True, exist_ok=True)
            path.write_text(tf["content"])

        # If the workspace is a CRA project, override to react-scripts test
        pkg = workspace / "package.json"
        if pkg.exists():
            import json as _json
            try:
                pkg_data = _json.loads(pkg.read_text())
                deps = {**pkg_data.get("dependencies", {}), **pkg_data.get("devDependencies", {})}
                if "react-scripts" in deps:
                    framework = "react-scripts"
            except Exception:
                pass

        cmd = FRAMEWORK_CMD.get(framework, ["python", "-m", "pytest", "-v"])
        env = {
            **__import__("os").environ,
            "CI": "true",
            "SKIP_PREFLIGHT_CHECK": "true",
            "NODE_OPTIONS": "--openssl-legacy-provider",
        }
        # For Node-based test frameworks, ensure deps are installed first
        if framework in ("vitest", "jest", "react-scripts") and pkg.exists():
            subprocess.run(
                ["npm", "install", "--prefer-offline", "--legacy-peer-deps"],
                cwd=workspace, capture_output=True, timeout=180,
            )
        proc = subprocess.run(cmd, capture_output=True, text=True, cwd=workspace, env=env)
        output = proc.stdout + proc.stderr
        success = proc.returncode == 0
        return AgentResult(success=success, output=output,
                           error=None if success else "tests_failed")

File: src/agents/testAgent.ts
import { BaseAgent, AgentResult } from "./base.js";
import { ModelTier } from "../router.js";

const SYSTEM = `You are a test engineer. Write tests for this project and make them pass.

You have tools available:
- bash_exec: run the test suite and see results
- read_file: read source files to understand what to test
- write_file: write test files
- list_dir: list directory contents

Workflow:
1. Use list_dir and read_file to understand the source code structure
2. Write tests using write_file — import only from files that actually exist
3. Run the tests with bash_exec to see results
4. Fix any failing tests by writing corrected files
5. Repeat until tests pass or you have exhausted reasonable fixes
6. Write a summary of what you tested and the final result

Critical rules:
- ONLY import from files that ACTUALLY EXIST (verify with read_file first)
- Do NOT invent utility functions that don't exist in the source
- Keep tests simple — test one behaviour per test`;

export class TestAgent extends BaseAgent {
  protected tier = ModelTier.REASONING;

  async run(args: Record<string, unknown>): Promise<AgentResult> {
    const architecture = args["architecture"] as string;
    let arch: Record<string, unknown> = {};
    try { arch = typeof architecture === "string" ? JSON.parse(architecture) : architecture; } catch {}
    const framework = String(arch["test_framework"] ?? "jest");
    const workspace = String(args["workspace"] ?? "");
    const messages: any[] = [
      { role: "system", content: SYSTEM },
      { role: "user", content: `Test framework: ${framework}\nWorkspace root: ${workspace}` },
    ];
    const summary = await this.runAgenticLoop(messages, workspace);
    const lower = summary.toLowerCase();
    const passed = lower.includes("pass") || summary.includes("✓") || lower.includes("success") || lower.includes("all tests");
    return { success: passed, output: summary, error: passed ? undefined : "tests_failed" };
  }
}
